Add new page for self-healing (#50212)
kubernetes/website: content/en/docs/concepts/architecture/self-healing.md

---
title: Kubernetes Self-Healing  
content_type: concept  
Weight: 50  
---
<!-- overview -->

Kubernetes is designed with self-healing capabilities that help maintain the health and availability of workloads. 
It automatically replaces failed containers, reschedules workloads when nodes become unavailable, and ensures that the desired state of the system is maintained.

<!-- body -->

## Self-Healing capabilities {#self-healing-capabilities} 

> - **Pod Restart:** If a container inside a Pod fails, Kubernetes restarts it based on the [`restartPolicy`](/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy)`. 

Reviewer: ```suggestion
- **Container-level restarts:** If a container inside a Pod fails, Kubernetes restarts it based on the [`restartPolicy`](/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy)`.
```

> - **Replica replacement:** If a Pod in a [`Deployment`](/docs/concepts/workloads/controllers/deployment/) or [`ReplicaSet`](/docs/concepts/workloads/controllers/replicaset/) fails, Kubernetes creates a replacement Pod to maintain the specified number of replicas. 

Reviewer: - **Replica replacement:** If a Pod in a [Deployment](/docs/concepts/workloads/controllers/deployment/) or [StatefulSet](/docs/concepts/workloads/controllers/statefule/) fails, Kubernetes creates a replacement Pod to maintain the specified number of replicas.
  If a Pod fails that is part of a [DaemonSet](/docs/concepts/workloads/controllers/daemonset/), the control plane
  creates a replacement Pod to run on the same node.
```

> - **Persistent Volume recovery:** If a node running a Pod with a Persistent Volume Claim (PVC) fails, Kubernetes can reattach the volume to a new Pod on a different node. 

Reviewer: ```suggestion
- **Persistent storage recovery:** If a node is running a Pod with a PersistentVolume (PV) attached, and the node fails, Kubernetes can reattach the volume to a new Pod on a different node. 
```

> - **Service Load Balancing:** If a Pod behind a [Service](/docs/concepts/services-networking/service/) fails, Kubernetes automatically removes it from the Service's endpoints to route traffic only to healthy Pods. 

Reviewer: ```suggestion
- **Load balancing for Services:** If a Pod behind a [Service](/docs/concepts/services-networking/service/) fails, Kubernetes automatically removes it from the Service's endpoints to route traffic only to healthy Pods.
```

## Components involved {#components} 

- **Kubelet:** Ensures that containers are running and restarts those that fail. 

> - **ReplicaSet controller:** Maintains the desired number of Pod replicas. 

Reviewer: ```suggestion
- **ReplicaSet, StatefulSet and DaemonSet controller:** Maintains the desired number of Pod replicas.
```

> - **Persistent Volume controller:** Manages volume attachment and detachment for stateful workloads. 

Reviewer: ```suggestion
- **PersistentVolume controllers:** Manages volume attachment and detachment for stateful workloads.
```

## Considerations {#considerations} 

- **Storage Failures:** If a persistent volume becomes unavailable, recovery steps may be required. 

- **Application Errors:** Kubernetes can restart containers, but underlying application issues must be addressed separately. 

## {{% heading "whatsnext" %}} 

- Read more about [Pods](/docs/concepts/workloads/pods/) 
- Learn about [Kubernetes Controllers](/docs/concepts/architecture/controller/) 

> - Explore [Persistent Volumes](/docs/concepts/storage/persistent-volumes/) 

Reviewer: ```suggestion
- Explore [PersistentVolumes](/docs/concepts/storage/persistent-volumes/) 
- Read about [node autoscaling](/docs/concepts/cluster-administration/node-autoscaling/). Node autoscaling
  also provides automatic healing if or when nodes fail in your cluster.
```

---
title: Kubernetes Self-Healing  
content_type: concept  
Weight: 50  
---
<!-- overview -->

Kubernetes is designed with self-healing capabilities that help maintain the health and availability of workloads. 
It automatically replaces failed containers, reschedules workloads when nodes become unavailable, and ensures that the desired state of the system is maintained.

<!-- body -->

## Self-Healing capabilities {#self-healing-capabilities} 

- **Container-level restarts:** If a container inside a Pod fails, Kubernetes restarts it based on the [`restartPolicy`](/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy)`.

> - **Replica replacement:** If a Pod in a [`Deployment`](/docs/concepts/workloads/controllers/deployment/) or [`ReplicaSet`](/docs/concepts/workloads/controllers/replicaset/) fails, Kubernetes creates a replacement Pod to maintain the specified number of replicas. If a Pod fails that is part of a [`DaemonSet`](/docs/concepts/workloads/controllers/daemonset/), the control plane creates a replacement Pod to run on the same node.

Reviewer: Oops, this is the suggestion I meant to make yesterday
```suggestion
- **Replica replacement:** If a Pod in a [Deployment](/docs/concepts/workloads/controllers/deployment/) or [StatefulSet](/docs/concepts/workloads/controllers/statefulset/) fails, Kubernetes creates a replacement Pod to maintain the specified number of replicas.
  If a Pod fails that is part of a [DaemonSet](/docs/concepts/workloads/controllers/daemonset/) fails, the control plane
  creates a replacement Pod to run on the same node.
```

- **Persistent storage recovery:** If a node is running a Pod with a PersistentVolume (PV) attached, and the node fails, Kubernetes can reattach the volume to a new Pod on a different node.

- **Load balancing for Services:** If a Pod behind a [Service](/docs/concepts/services-networking/service/) fails, Kubernetes automatically removes it from the Service's endpoints to route traffic only to healthy Pods.

## Components involved {#components} 

> 

Reviewer: ```suggestion

Here are some of the key components that provide Kubernetes self-healing:

```

> - **Kubelet:** Ensures that containers are running and restarts those that fail.

Reviewer: (optional)
```suggestion
- **[kubelet](/docs/concepts/architecture/#kubelet):** Ensures that containers are running, and restarts those that fail.
```

## Self-Healing Capabilities {#self-healing-capabilities} 

- **Pod Restart:** If a container inside a Pod fails, Kubernetes restarts it based on the [`restartPolicy`](/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy)`. 

- **Replica Replacement:** If a Pod in a [`Deployment`](/docs/concepts/workloads/controllers/deployment/) or [`ReplicaSet`](/docs/concepts/workloads/controllers/replicaset/) fails, Kubernetes creates a replacement Pod to maintain the specified number of replicas. 

- **Node Failover:** If a node becomes unreachable, Kubernetes reschedules affected Pods onto available nodes. 

- **Persistent Volume Recovery:** If a node running a Pod with a Persistent Volume Claim (PVC) fails, Kubernetes can reattach the volume to a new Pod on a different node. 

- **Service Load Balancing:** If a Pod behind a [Service](/docs/concepts/services-networking/service/) fails, Kubernetes automatically removes it from the Service's endpoints to route traffic only to healthy Pods. 

## Components Involved {#components} 

- **Kubelet:** Ensures that containers are running and restarts those that fail. 

- **ReplicaSet Controller:** Maintains the desired number of Pod replicas. 

> - **Node Controller:** Detects node failures and reschedules affected Pods. 

Reviewer: You are wrong here. Check your understanding of what manages nodes in a Kubernetes cluster.

Author: thank you for pointing it out, you are right.
I have removed the line for node controller as it doesn't actively contribute to self-healing
